### ISSUE TITLE
mini.pick not modifier

### ISSUE THREAD

**Issue by ibutra** [echasnovski/mini.nvim#575]
### Contributing guidelines

- [X] I have read [CONTRIBUTING.md](https://github.com/echasnovski/mini.nvim/blob/main/CONTRIBUTING.md)
- [X] I have read [CODE_OF_CONDUCT.md](https://github.com/echasnovski/mini.nvim/blob/main/CODE_OF_CONDUCT.md)

### Module(s)

mini.pick

### Description

Hi,

I am really enjoying mini.pick but would like to ask a question:

Would you be opposed to introduce a not operator/modifier like fzf has?
Basically like the `'` for an exact match, but all entries that do match are filtered out.
I would propose to also use `!` for this. 

If you are ok with this would you rather implement this yourself or are you open to a pull request for this?

Greetings
Stefan

**Comment by echasnovski** [comment]
Thanks for the suggestion!

I remember thinking about it and still not implementing. Here are the few reasons why:
- It is not quite intuitive for me in the way you describe and quite complex for the way I find intuitive. Having `!abc` prompt match items not containing `abc` exactly is a two-step jump from having `abc` match items fuzzily. So I'd expect `!abc` to match items which don't match `abc` *fuzzily*. But that would require allowing extra modifies to play nicely with `!`: `!'abc` would mean "not matches exactly", `!^abc` - "not matches exactly at start", `!abc$` - not matches at the end, and `!*abc` not matches fuzzily (useful if user wants something like `!*'a`).
- It adds complexity to a parsing logic. Which is already quite crowded.

I'll revisit it later, maybe I'd reevaluate.

**Comment by ibutra** [comment]
Oh I would be totally fine with your way, I actually think it is more consistent. What I like about it is, that it should make the implementation actually quite straight forward in that instead of filtering non matches out you only keep the non matching.

In the simplest form though not the most performant we could simply remove the matches of the current methods from the original list.

**Comment by echasnovski** [comment]
> If you still don't like it that is also totally fine for me, I don't want to push you into something you don't like. In that case I will just implement it for myself on top of `MiniPick.default_match`

No worries. For the time being, that is the best approach for you. It'll take time for this to come to life (need some time to reevaluate).

**Comment by ibutra** [comment]
Just in case somebody wants this now here is my implementation:
[Gist](https://gist.github.com/ibutra/24a741496dd60bb2e33a045d84db42b1)

This should be an equivalent algorithm to the one used in Mini.Pick but with added `!` operator.
Note: The original sorting is preserved with the `!` operator because obviously I can't sort based on the length of the match.

Also anybody including you echanovski feel free to use this in any way you like including inspiration ;)

**Comment by echasnovski** [comment]
> Just in case somebody wants this now here is my implementation: [Gist](https://gist.github.com/ibutra/24a741496dd60bb2e33a045d84db42b1)
> 
> This should be an equivalent algorithm to the one used in Mini.Pick but with added `!` operator. Note: The original sorting is preserved with the `!` operator because obviously I can't sort based on the length of the match.
> 
> Also anybody including you echanovski feel free to use this in any way you like including inspiration ;)
> 
> Also any feedback is of course welcome

That's quite a bottoms up implementation. I was thinking more like utilizing [`default_match()`](https://github.com/echasnovski/mini.nvim/blob/05f4a49cd85a67b90328a1bcbae4d9ed2a0a417b/doc/mini-pick.txt#L834) with `do_sync` argument to return the result. Like this:
```lua
local default_match = require('mini.pick').default_match
local match_ext = function(stritems, inds, query)
  if query[1] ~= '!' then return default_match(stritems, inds, query) end

  -- Compute matches ignoring first '!'
  local new_query = vim.list_slice(query, 2, #query)
  local not_matches = default_match(stritems, inds, new_query, true)

  -- Compute and return complement of the matched indexes
  local res_inds_map = {}
  for i = 1, #stritems do
    res_inds_map[i] = true
  end
  for _, ind in ipairs(not_matches) do
    res_inds_map[ind] = nil
  end
  return vim.tbl_keys(res_inds_map)
end
```

Its only downside is that it is synchronous when query starts with '!' (should be still asynchronous when it doesn't).

**Comment by ibutra** [comment]
That was my first thought but I missed the 'do_sync' option as I didn’t update mini before and I wanted to keep the asynchronous functionality because I have a very large file list I need to search.

**Comment by echasnovski** [comment]
After rethinking this through and looking again at code, I still think that implementing this would add too much complexity to the already crowded logic of query parsing.

So sadly I'll close this as not planned.
